avoid materialising meta (baggage + tags) during serialisation

richardstartin opened the pull request:
Creating the meta map is a minor CPU bottleneck in serialisation, this change avoids allocating a new map or any map resizing.

richardstartin commented on dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java:
```diff
@@ -119,10 +110,13 @@ public void writeDDSpan(final DDSpan span, final DEST destination) throws IOExce
   }
 
+  /** A destination to serialize the span's meta into */
+  protected abstract MapAcceptor<String> getMetaAcceptor(DEST destination);
```
I haven't measured it, but I expect this to get hot and there will only be implementation at runtime, so the `MapAcceptor` might not actually get allocated, it's still smaller than a hashmap and doesn't need resizing.

bantonsson replied:
Yeah, it would be interesting to see, but I think it is a win either way.

richardstartin commented on dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java:
```diff
@@ -40,6 +41,12 @@ public void writeString(byte[] key, String value, MessagePacker destination) thr
   }
 
+  @Override
+  public void writeTag(byte[] key, String value, MessagePacker destination) throws IOException {
```
This got backed out in a bad merge (my bad) :(

richardstartin commented on dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:
```diff
@@ -354,7 +356,7 @@ public String getSpanType() {
   @Override
   public Map<String, Object> getTags() {
-    return context.getTags();
+    return Collections.unmodifiableMap(context.getTags());
```
Move the wrapping here so it can be used without wrapping.

richardstartin commented on dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:
```diff
@@ -349,8 +349,8 @@ public synchronized void setTag(final String tag, final Object value) {
     }
   }
 
-  public synchronized Map<String, Object> getTags() {
```
Wasn't sure what `synchronized` was achieving here. `Collections.unmodifiableMap` doesn't mutate the map, and prevented it from being mutated. The only caller of this method, `DDSpan` does not mutate the map. Should the visibility be reduced to package?

devinsba replied:
This is a new module. I would make it package private if possible

richardstartin replied:
Right, but it doesn't actually make any concurrency guarantees. If the map can only be mutated within critical sections, it will now have been released and any caller can observe the map during critical sections where the map is being mutated. If the map were being _copied_ at this point, `synchronized` would make sense.

bantonsson commented on dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:
```diff
@@ -261,19 +261,21 @@ public final DDSpan setSpanType(final String type) {
-    final Map<String, String> meta = new HashMap<>();
-    for (final Map.Entry<String, String> entry : context.getBaggageItems().entrySet()) {
-      meta.put(entry.getKey(), entry.getValue());
+  public void transferMeta(MapAcceptor<String> acceptor) {
```
Yes, the `getMeta` show up in the memory overhead as well, and is creating a lot of unnecessary `HashMapNode` and `HashMapNode[]`.

tylerbenson commented on dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java:
```diff
@@ -41,50 +41,49 @@ public void writeString(final byte[] key, final String value, final MessagePacke
   }
 
   @Override
-  public void writeFloat(final byte[] key, final float value, final MessagePacker destination)
-      throws IOException {
+  public void writeFloat(byte[] key, float value, MessagePacker destination) throws IOException {
     writeKey(key, destination);
```
As determined by our Idea settings, these finals get automatically added when we save the file: https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md
I'd suggest not trying to change them back, otherwise the next time it gets saved will redo the change.

richardstartin replied:
What I did was rebase on top of your change made to this file after I submitted this PR. I accepted my changes over yours. If you want, I can put the final parameters back in, but would prefer that my editor doesn't make changes without my control.

devinsba replied:
Might need to rebase against master? The version is dependent on the most recent tag in the history

tylerbenson commented on .palantir/revapi.yml:
```diff
@@ -299,6 +299,9 @@ acceptedBreaks:
       old: "method datadog.trace.bootstrap.instrumentation.api.AgentScope datadog.trace.core.scopemanager.DDScopeManager::activate(datadog.trace.bootstrap.instrumentation.api.AgentSpan,\
         \ boolean)"
       justification: "Removed deprecated APIs preventing refactoring"
+    - code: "java.method.removed"
+      old: "method java.util.Map<java.lang.String, java.lang.String> datadog.trace.core.DDSpan::getMeta()"
+      justification: "internal api"
```
Odd... why aren't these added to the `0.52.0` section below?

richardstartin replied:
see the comment about revapi below. Happy to learn what I did wrong.

devinsba replied:
Might need to rebase against master? The version is dependent on the most recent tag in the history

richardstartin replied:
No, thanks for the suggestion, I even tried it, it produces the same result. I had also rebased against `6f549f3bac29c674a5127624836c3b918c22392e` btw.

randomanderson replied:
@devinsba is probably right.  Originally, this branch was against 0.51.0, you added the revapi changes, then you rebased to the latest after 0.51.0 was released.  The commands you used weren't wrong.

richardstartin replied:
I created this branch last night off the tip of master. 0.51.0 was released 11 days ago. Does it sound likely that I branched prior to 0.51.0?

randomanderson replied:
@richardstartin Maybe you didn't fetch all tags?  `./gradlew currentVersion` will tell you what gradle thinks the current version of the project is

richardstartin replied:
```
> Task :currentVersion

Project version: 0.52.0-SNAPSHOT
```

Of course, I can't verify what it was at 23:16 UTC last night.

randomanderson replied:
Idk then.  The commands you ran looked fine

richardstartin replied:
The minimal reproducer is this: 

```
git checkout master
git fetch # to be sure
git pull origin master
git checkout -b revapi-bug
```

change the name of `DDSpan.getMeta` to `thisIsABreakingChange`

Run 

```
./gradlew :dd-trace-ot:revapiAcceptBreak --justification "this is a breaking change for test purposes" \
          --code "java.method.removed" \
          --old "method java.util.Map<java.lang.String, java.lang.String> datadog.trace.core.DDSpan::getMeta()"
```

The change is attributed to the wrong version according to our expectations (should create a new section 0.52.0):
```git diff

diff --git a/.palantir/revapi.yml b/.palantir/revapi.yml
index e18bbf39f..7d290c945 100644
--- a/.palantir/revapi.yml
+++ b/.palantir/revapi.yml
@@ -299,6 +299,9 @@ acceptedBreaks:
       old: "method datadog.trace.bootstrap.instrumentation.api.AgentScope datadog.trace.core.scopemanager.DDScopeManager::activate(datadog.trace.bootstrap.instrumentation.api.AgentSpan,\
         \ boolean)"
       justification: "Removed deprecated APIs preventing refactoring"
+    - code: "java.method.removed"
+      old: "method java.util.Map<java.lang.String, java.lang.String> datadog.trace.core.DDSpan::getMeta()"
+      justification: "this is a breaking change for test purposes"
     - code: "java.method.removed"
       old: "method long datadog.trace.common.writer.ddagent.AbstractDisruptor<T>::getCurrentCount()\
         \ @ datadog.trace.common.writer.ddagent.TraceProcessingDisruptor"
diff --git a/dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java b/dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java
index 9bf2ce56a..3d659b745 100644
--- a/dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java
+++ b/dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java
@@ -265,7 +265,7 @@ public class DDSpan implements MutableSpan, AgentSpan {
    *
    * @return merged context baggage and tags
    */
-  public Map<String, String> getMeta() {
+  public Map<String, String> thisIsABreakingChange() {
     final Map<String, String> meta = new HashMap<>();
     for (final Map.Entry<String, String> entry : context.getBaggageItems().entrySet()) {
       meta.put(entry.getKey(), entry.getValue());
diff --git a/dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java b/dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java
index 77307a838..670d11855 100644
--- a/dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java
+++ b/dd-trace-core/src/main/java/datadog/trace/core/serialization/FormatWriter.java
@@ -119,7 +119,7 @@ public abstract class FormatWriter<DEST> {
     /* 9  */ writeTag(TYPE, span.getType(), destination);
     /* 10 */ writeInt(ERROR, span.getError(), destination);
     /* 11 */ writeNumberMap(METRICS, span.getMetrics(), destination);
-    /* 12 */ writeStringMap(META, span.getMeta(), destination);
+    /* 12 */ writeStringMap(META, span.thisIsABreakingChange(), destination);
     writeMapFooter(destination);
   }
 
```

richardstartin commented:
On revapi, here's what I did:

```
./gradlew :dd-trace-ot:revapiAcceptBreak --justification "internal api" \
          --code "java.method.removed" \
          --old "method java.util.Map<java.lang.String, java.lang.String> datadog.trace.core.DDSpan::getMeta()"

./gradlew :dd-trace-ot:revapiAcceptBreak --justification "internal api" \
          --code "java.method.removed" \
          --old "method void datadog.trace.core.serialization.FormatWriter<DEST>::writeStringMap(java.lang.String, java.util.Map<java.lang.String, java.lang.String>, DEST) throws java.io.IOException"

./gradlew :dd-trace-ot:revapiAcceptBreak --justification "internal api" \
          --code "java.method.removed" \
          --old "method void datadog.trace.core.serialization.FormatWriter<DEST>::writeStringMap(java.lang.String, java.util.Map<java.lang.String, java.lang.String>, DEST) throws java.io.IOException @ datadog.trace.core.serialization.JsonFormatWriter"

./gradlew :dd-trace-ot:revapiAcceptBreak --justification "internal api" \
          --code "java.method.removed" \
          --old "method void datadog.trace.core.serialization.FormatWriter<DEST>::writeStringMap(java.lang.String, java.util.Map<java.lang.String, java.lang.String>, DEST) throws java.io.IOException @ datadog.trace.core.serialization.JsonFormatWriter"
```

What did I do wrong?
